### ISSUE TITLE
[REQUEST] Raylib Tileson

### ISSUE THREAD

**Issue by Nathan-MV** [vaiorabbit/raylib-bindings#5]
So we can use Tiled maps
https://github.com/RobLoach/raylib-tileson/

**Comment by vaiorabbit** [comment]
https://github.com/vaiorabbit/raylib-bindings-tileson
* Note that this repo is work-in-progress. I'm sure codes there would never work at all.

**Comment by vaiorabbit** [comment]
Added [`examples/raylib_with_tileson.rb`](https://github.com/vaiorabbit/raylib-bindings/blob/7a181563bec27b2724b01b830baf2d3c411cd410/examples/raylib_with_tileson.rb). For now you can test it with pre-release version of `raylib-bindings-tileson` gem:

```
$ pwd
    /your/home/raylib-bindings
$ gem install raylib-bindings-tileson --pre
    Fetching raylib-bindings-tileson-0.6.0pre-arm64-darwin.gem
    Fetching raylib-bindings-0.6.0-arm64-darwin.gem
    Successfully installed raylib-bindings-0.6.0-arm64-darwin
    Successfully installed raylib-bindings-tileson-0.6.0pre-arm64-darwin
    2 gems installed
$ cd examples
$ ruby raylib_with_tileson.rb
```

**Comment by Nathan-MV** [comment]
Strange, i did it the same way i did with raylib-bindings, but it can't load tileson.dll

![image](https://github.com/vaiorabbit/raylib-bindings/assets/67521199/2fb5de0a-7f22-4c91-97bc-9a2f3e72bca2)

![image](https://github.com/vaiorabbit/raylib-bindings/assets/67521199/e999d620-837d-48d0-87c4-ecadcf844661)


```
Could not open library 'D:/Cloud/Projects/Raylib/Ruby/rgray/binaries/lib/ruby/gems/3.3.0/gems/tileson.dll': Failed with error 126: The specified module could not be found.
.
Searched in <system library path>, /usr/lib, /usr/local/lib, /opt/local/lib
```

**Comment by vaiorabbit** [comment]
It seems all the libraries bundled with Gem packages are moved from their default location. Maybe your script is failing to find the DLLs. Please consider adding the following code to add a path to make your script look for the DLLs.

```ruby
if System.windows?
  if Module.const_defined?('RubyInstaller')
    RubyInstaller::Runtime.add_dll_directory(File.expand_path(game_path))
  end
  # ...
```

Ref.: https://github.com/vaiorabbit/sdl2-bindings-whacamole/blob/main/main.rb#L22-L25C8


`tileson.dll` in `version 0.6.0pre` is built with `libraylib.dll` in `raylib-bindings 0.6.0`. Trying to use `tileson.dll` with newer DLLs (such as `libraylib.dll` uploaded to GitHub) or with older DLLs (such as `libraylib.dll` in `v0.5.0`) may cause problems.

**Comment by Nathan-MV** [comment]
`raylib-bindings 0.6.0` loads just fine, but when i try to use something from tileson it crashs with
``D:/Cloud/Projects/Raylib/Ruby/rgray/lib/node/gameplay.rb:12:in `draw': undefined method `DrawTiled' for module Raylib (NoMethodError)``

and in debug mode it gives the same warning 
`Could not open library 'D:/Cloud/Projects/Raylib/Ruby/rgray/binaries/lib/ruby/gems/3.3.0/gems/tileson.dll': Failed with error 126: The specified module could not be found.`

 I disabled gems to supposedly start ruby faster with 
``"./binaries/ruby.exe" --disable=gems,rubyopt,did_you_mean game.rb %*``
So basically, i'm copying and pasting ruby.exe and the lib folder into the binaries folder, as idk how to create an executable for ruby 3.3.0 without embedding it into C like Taylor did with MRuby https://github.com/HellRok/Taylor

**Comment by Nathan-MV** [comment]
I've made some minor adjustments to the code, but it still gives the same error for tileson

```# Set the environment variable 'GAME_PATH' to the directory containing the script
ENV['GAME_PATH'] = File.dirname(File.expand_path(__FILE__))

# The Game module encapsulates functionality related to the game.
module Game
  GAME_PATH = ENV['GAME_PATH'] || '.'
  GEMS_PATH = File.join(GAME_PATH, 'binaries', 'lib', 'ruby', 'gems', RUBY_VERSION, 'gems')
  SCRIPT_DIR = File.join(GAME_PATH, 'lib')
  
  module_function
  
  # @return [void]
  def load_libraries
    # Add the gem path to the LOAD_PATH
    $LOAD_PATH.unshift(GEMS_PATH)

    # Require necessary gems
    %w[active_support benchmark csv colorize fileutils ffi json json-schema logger tileson parallel raylib yaml zeitwerk
       zlib].each do |lib|
      require lib
    end

    $LOAD_PATH.unshift(File.join(GAME_PATH, 'ext', 'ruby'))
    require "#{GAME_PATH}/ext/ruby/logger"
    require "#{GAME_PATH}/ext/ruby/yaml"
    require 'system'

    System.x86_64? ? arch = 'x86_64' : arch = 'arm64'
    if System.windows?
      RubyInstaller::Runtime.add_dll_directory(GEMS_PATH) if Module.const_defined?
      Raylib.load_lib("#{GEMS_PATH}/libraylib.dll",
                      raygui_libpath: "#{GEMS_PATH}/raygui.dll",
                      physac_libpath: "#{GEMS_PATH}/physac.dll")
      Raylib.load_tileson_lib("#{GEMS_PATH}/tileson.dll")
    elsif System.mac?
      Raylib.load_lib("#{GEMS_PATH}/libraylib.#{arch}.dylib",
                      raygui_libpath: "#{GEMS_PATH}/raygui.#{arch}.dylib",
                      physac_libpath: "#{GEMS_PATH}/physac.#{arch}.dylib")
      Raylib.load_tileson_lib("#{GEMS_PATH}/tileson.#{arch}.dylib")
    elsif System.linux?
      Raylib.load_lib("#{GEMS_PATH}/libraylib.#{arch}.so",
                      raygui_libpath: "#{GEMS_PATH}/raygui.#{arch}.so",
                      physac_libpath: "#{GEMS_PATH}/physac.#{arch}.so")
      Raylib.load_tileson_lib("#{GEMS_PATH}/tileson.#{arch}.so")
    else
      Logger.error("Unknown OS: #{RUBY_PLATFORM}")
    end
  end

**Comment by vaiorabbit** [comment]
Then please search `libstdc++-6.dll` from your MSys2 installation and and copy it into your directory (e.g: working directory you launch the command `"./binaries/ruby.exe"`).

(maybe you'll also need to copy `libgcc_s_seh-1.dll`)

```
> dir /w
...
[.]                   [..]                  [binaries]            game.rb               libstdc++-6.dll
[map]
...

> .\binaries\ruby --disable=gems,rubyopt,did_you_mean game.rb
```

These DLLs are sometimes the source of dependency problems for C++ programs built on MinGW: https://github.com/search?q=libstdc%2B%2B-6.dll&type=issues

**Comment by Nathan-MV** [comment]
Tileson doesn't even work the normal way with `ruby game.rb` and
```
  def setup_raylib_windows(_arch)
    load_raylib_lib("#{Gem::Specification.find_by_name('raylib-bindings').full_gem_path}/lib/libraylib.dll",
                    "#{Gem::Specification.find_by_name('raylib-bindings').full_gem_path}/lib/raygui.dll",
                    "#{Gem::Specification.find_by_name('raylib-bindings').full_gem_path}/lib/physac.dll",
                    "#{Gem::Specification.find_by_name('raylib-bindings-tileson').full_gem_path}/lib/tileson.dll")
  end
  
  def load_raylib_lib(raylib_path, raygui_path, physac_path, tileson_path)
    Raylib.load_lib(raylib_path, raygui_libpath: raygui_path, physac_libpath: physac_path)
    Raylib.load_tileson_lib(tileson_path)
  end
```

**Comment by Nathan-MV** [comment]
I managed to run it with `ruby game.rb`, i was putting LoadTiled in the load method, as i was doing with the textures and images, so it wasn't working for some reason , and `libstdc++-6.dll` solved it for the `"./binaries/ruby.exe"`
![image](https://github.com/vaiorabbit/raylib-bindings/assets/67521199/60b710fa-1923-4cbb-a1c5-a8682cd590a1)
